src/analysis/souffle/taint_with_auth_logic.dl

```diff
@@ -0,0 +1,121 @@
+// it).
+.decl ownsTag(owner: Principal, tag: Tag)
+
+// Some accessPath is actually downgraded (by taking into consideration the
+// privileges of a party trying to downgrade it, and as a result of that party
+// deciding to downgrade it). Because here IFC labels are sets of tags (ordered
+// by the subset or equals relation), the downgrade works by removing a tag
+// from the set (making it "lower" in the order).
+.decl downgrades(path: AccessPath, tag: Tag)
+
+// A Tag is a member of the IFCLabel ell (which is a set of Tags).
+.decl isMember(tag: Tag, ell: IFCLabel)
+
+// The IFCLabel of an accessPath flowsTo another IFCLabel.
+// This is equivalent to "subset or equals". The final outputs (i.e. egress
+// points) can be checked by doing a "accessPathIFCLabelFlowsTo" check to a
+// particular expected IFCLabel.
+.decl accessPathLabelFlowsTo(path: AccessPath, ell: IFCLabel)
+
+// Because datalog does not have universal quantifiers, accessPathLabelFlowsTo 
+// (which is "subset or equal to") is defined by proving that there is no 
+// counterexample to the subset relation. A counterexample to the relation 
+// `isSubsetOrEqual(Set1, Set2)` is an element that is in Set1 but not Set2. 
+// The relation `accessPathLabelDoesNotFlowTo` is true when there is a
+// counterexample, and accessPathLabelFlowsTo is just the negation of
+// accessPathLabelDoesNotFlowTo.
+.decl accessPathLabelDoesNotFlowTo(path: AccessPath, ell: IFCLabel)
+
+// This is a variation on the `mayHaveTag` relation defined in `taint.dl`
+// It says that an accessPath may have been tainted with tag. By contrast to 
+// the other implementation, this one takes into account downgrades (which 
+// remove tags). Downgrades happen as a result of statements made in 
+// authorization logic.
+.decl mayHaveTagAL(accessPath: AccessPath, tag: Tag)
+
+//-----------------------------------------------------------------------------
+// Universe Relations
+//-----------------------------------------------------------------------------
+// These relations say that some object is a member of the universe for that
+// object. These are useful for scoping when there is negation or otherwise to 
+// ensure that variables are grounded.
+.decl isIFCLabel(ell: IFCLabel)
+
+//-----------------------------------------------------------------------------
+// Rules
+//-----------------------------------------------------------------------------
+downgrades(path, tag) :- ownsTag(owner, tag), saysDowngrades(owner, path, tag).
+
+// The following definition of mayHaveTagAL would not work, because when a tag 
+// is downgraded from an accessPath it should also not propagate to other paths 
+// that are after it in the graph. The following incorrect implementation does 
+// not have that behavior
+// mayHaveTagAL(path, tag) :- mayHaveTag(path, tag), !downgrades(path, tag).
+
+// Real implementation of mayHaveTagAL
+
+mayHaveTagAL(tgt, tag) :- claimHasTag(tgt, tag).
```

Comment:
Is the idea to keep this separate from `mayHaveTag` until we figure it out fully? Ideally, this should be `mayHaveTag`. OK to keep them separate for now, but we should merge them eventually.